### ISSUE TITLE
Implement value function iteration

### ISSUE THREAD

**Issue by albop** [EconForge/dolo#22]

**Comment by yaniv256** [comment]
I think I'd like to give this a try. Any pointers?

**Comment by albop** [comment]
Excellent !

I will be very interested to follow your progress. Here are some pointers and remarks.
I just pushed a few refactoring commits, reorganizing models in a more consistent ways (note that dolo.algos.fg and dolo.algos.mfg have been renamed to `dolo.algos.dtcscc` and `dolo.algos.dtmscc` respectively). The models are classified according to the hierarchy here: https://github.com/EconForge/heterogeneous_agents_workshop/wiki/Model-classification-%28DRIMS%29 . The idea is that a given type of models (e.g. dtcscc) can have several optional functions including arbitrage and value-updating functions. This was exactly meant to allow for concurrent solution methods with the same model.

You probably want to implement your function with continuous states first, since it is the type of model you have used with the rbc notebook. Reproducing the same steps with mixed states (i.e. continuous + discrete markov states) will be relatively trivial.

The way I see it, there are to milestones for the vfi implementation:
1/ get a proof of concept algorithm to work, without focusing on performance. It should be relatively easy using one of the constrained optimization algorithm from `scipy.optimize`.
2/ vectorize or jit this solution method. This is crucial since iterating over all the points will have a big performance cost. I now have a clearer idea about how to do that. One crucial element is to have a simple implementation of an optimization routine, that we can try to compile. That could involve a bit of  numba voodoo, but at least in principle, it should be feasible.

I'll be happy to help at any stage. The complications from stage 2/ are not completely predictible and may involve waiting (for numba to add some features).

To help with the implementation itself I have compiled a gist, with a few instructions to recover the model functions (https://gist.github.com/albop/ab49de476b6bcbdd689c) and the model calibration. Note that the definition of the value updating equation currently maches the `value_2` function of the model classification.
If you want you can also look at the `dolo.algos.dtcscc.vfi` file. The available functions compute the value matching a given decision rule. What you need to implement is a similar function, but with an improving step, the main difference being that you need to iterate over the points, rather than evaluating the grid at once.

Again, I would be very glad to see that happen.

**Comment by yaniv256** [comment]
Thank you so so much.

I had a bit of a letdown with numba when I tested it for a scipy.optimize.fsolve parameter space swipe. I got it to work but the execution time dropped much less than what I was hoping for (less than 20% I believe). I'm not sure I gave numba enough static typing to really test it, but my interim conclusion was that since fsolve probably runs on Fortran, and that is where cpu is spent, I save practically nothing with JIT. I dumped numba and moved on to think that the real source for performance gains must be information sharing across optimizations (besides parallelization, of course). The obvious thing is doing some learning to generate improved initial guesses. For my fsolve problem it was enough to wrap fsolve inside a class that remembered the last fsolve output, if it was good. Just thought I'd share that.

How does your numba experience compare?
